src/video_colorspace.cpp

```diff
@@ -0,0 +1,174 @@
+#include "video_colorspace.h"
+
+extern "C" {
+#include <libswscale/swscale.h>
+}
+
+namespace video {
+
+  bool
+  colorspace_is_dynamic(colorspace_e colorspace) {
+    return colorspace == colorspace_e::bt2020sdr || colorspace == colorspace_e::bt2020;
+  }
+
+  bool
+  colorspace_is_dynamic(const sunshine_colorspace_t &colorspace) {
+    return colorspace_is_dynamic(colorspace.colorspace);
+  }
+
+  bool
+  colorspace_needs_dynamic_metadata(colorspace_e colorspace) {
+    return colorspace == colorspace_e::bt2020;
+  }
+
+  bool
+  colorspace_needs_dynamic_metadata(const sunshine_colorspace_t &colorspace) {
+    return colorspace_needs_dynamic_metadata(colorspace.colorspace);
+  }
+
+  sunshine_colorspace_t
+  colorspace_from_rtsp_csc(int csc, bool hdr_requested, bool hdr_display) {
+    /* See video::config_t::encoderCscMode for details */
+
+    sunshine_colorspace_t colorspace;
+
+    // When HDR is requested, that overrides the SDR colorspace selection
+    if (hdr_requested) {
+      if (hdr_display) {
+        // Rec. 2020 with ST 2084 perceptual quantizer
+        colorspace.colorspace = colorspace_e::bt2020;
+      }
+      else {
+        // Rec. 2020
+        colorspace.colorspace = colorspace_e::bt2020sdr;

```

Comment:
Moonlight expects the host to encode using its requested colorspace even when streaming in 10-bit modes, so the above case should be:
```
      if (hdr_requested && hdr_display) {
        // Rec. 2020 with ST 2084 perceptual quantizer
        colorspace.colorspace = colorspace_e::bt2020;
      }
      else {
      ...

Reply:
Rec. 601 is probably correct. We generally request that as a default for historical reasons, except for some GPUs that only behave properly with Rec. 709. I'll take a look at the crash.

Btw, GFE defaults to Rec. 601 if the colorspace isn't recognized. It's just academic because Moonlight will not pass an invalid color space, but just mentioning it since you added a codepath for that.

Reply:
There was some misunderstanding on how 10-bit encoding is activated.
Pushed a fix https://github.com/LizardByte/Sunshine/pull/1427/commits/93d4b27dfdb934bc6114a163c873a27cda16456f